## **1. 라이브러리 설치**

In [ ]:
!pip install koreanize-matplotlib

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.9/7.9 MB 52.5 MB/s eta 0:00:00


In [ ]:
!pip install ultralytics

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 19.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 2.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 120.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 94.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 53.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 1.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 10.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 38.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 16.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 4.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 103.9 MB/s eta 0:00:00
  Attempting uninstall: nvidia-nvjitlink-cu12
    Found existing installation: nvidia-nvjitlink-cu12 12.5.82
    Uninstall

In [ ]:
from google.colab import files
import os
import zipfile
import json
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import cv2
import yaml
import shutil
from tqdm import tqdm
from glob import glob
import random
import numpy as np
import torch
from ultralytics import YOLO
from IPython.display import Image, display
import glob
import koreanize_matplotlib

Creating new Ultralytics Settings v0.0.6 file ✅ 
View Ultralytics Settings with 'yolo settings' or at '/root/.config/Ultralytics/settings.json'
Update Settings with 'yolo settings key=value', i.e. 'yolo settings runs_dir=path/to/dir'. For help see https://docs.ultralytics.com/quickstart/#ultralytics-settings.


## **2. 시드 고정**

In [ ]:
random.seed(71)
np.random.seed(71)
torch.manual_seed(71)
torch.cuda.manual_seed(71)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False

## **3. 데이터 경로 설정**

In [ ]:
# 데이터 경로
data_path = '/content/drive/MyDrive/codeit_level1_project/data'

# 훈련, 검증, 테스트 데이터 경로 설정
train_dir = os.path.join(data_path, 'train_images')
test_dir = os.path.join(data_path, 'test_images')
annotation_dir = os.path.join(data_path, 'train_annotations')

## **4. 하이퍼파라미터 조정**
- 본 실험에서는 최종 훈련 모델은 YOLOv8x로 설정하되,
튜닝 시간 단축과 자원 절약을 위해 YOLOv8n 모델로 하이퍼파라미터 튜닝을 진행하였다.
- 튜닝에는 Ultralytics에서 제공하는 model.tune() 기능을 사용하였으며, 총 50회 반복 탐색(iterations=50)을 수행하였다.

| 분류             | 파라미터                                                                         | 이유                             |
| -------------- | ---------------------------------------------------------------------------- | ------------------------------ |
| 학습률 최적화     | `lr0`, `lrf`, `momentum`, `weight_decay`, `warmup_epochs`, `warmup_momentum` | 학습 속도 및 초기 안정성 개선                  |
| 손실 함수 가중치   | `box`, `cls`                                                                 | 바운딩 박스 회귀와 클래스 분류 간 손실 균형 조        |
| 데이터 증강       | `hsv_h`, `hsv_s`, `hsv_v`, `fliplr`, `mosaic`                                | 다양한 시각적 변형을 통해 모델 일반화 능력 향상            |


In [ ]:
model = YOLO("yolov8n.pt")

search_space = {
    "lr0": (1e-5, 1e-1),
    "lrf": (0.01, 1.0),
    "momentum": (0.6, 0.98),
    "weight_decay": (0.0, 0.001),
    "warmup_epochs": (0.0, 5.0),
    "warmup_momentum": (0.0, 0.95),
    "box": (0.02, 0.2),
    "cls": (0.2, 4.0),
    "hsv_h": (0.0, 0.1),
    "hsv_s": (0.0, 0.9),
    "hsv_v": (0.0, 0.9),
    "fliplr": (0.0, 1.0),
    "mosaic": (0.0, 1.0)
}

model.tune(
    data="/content/drive/MyDrive/codeit_level1_project/data/data.yaml",
    epochs=40,
    iterations=50,
    imgsz=640,
    optimizer="AdamW",
    space=search_space,
    save=True,
    val=True,
    plots=True,
    project="/content/drive/MyDrive/codeit_level1_project/YOLOv8_epoch100_hyperparameters/yolo_tuning_results",
    name="yolo8_hyp_0534"

)

100%|██████████| 6.25M/6.25M [00:00<00:00, 390MB/s]

Tuner: Initialized Tuner instance with 'tune_dir=/content/drive/MyDrive/codeit_level1_project/YOLOv8_epoch100_hyperparameters/yolo_tuning_results/yolo8_hyp_0534'
Tuner: 💡 Learn about tuning at https://docs.ultralytics.com/guides/hyperparameter-tuning
Tuner: Starting iteration 1/50 with hyperparameters: {'lr0': 0.01, 'lrf': 0.01, 'momentum': 0.937, 'weight_decay': 0.0005, 'warmup_epochs': 3.0, 'warmup_momentum': 0.8, 'box': 0.2, 'cls': 0.5, 'hsv_h': 0.015, 'hsv_s': 0.7, 'hsv_v': 0.4, 'fliplr': 0.5, 'mosaic': 1.0}


Saved /content/drive/MyDrive/codeit_level1_project/YOLOv8_epoch100_hyperparameters/yolo_tuning_results/yolo8_hyp_0534/tune_scatter_plots.png
Saved /content/drive/MyDrive/codeit_level1_project/YOLOv8_epoch100_hyperparameters/yolo_tuning_results/yolo8_hyp_0534/tune_fitness.png

Tuner: 1/50 iterations complete ✅ (589.97s)
Tuner: Results saved to /content/drive/MyDrive/codeit_level1_project/YOLOv8_epoch100_hyperparameters/yolo_tuning_results/yolo8_hyp_0534
Tuner: Best fitness=0.84648 observed at iteration 1
Tuner: Best fitness metrics are {'metrics/precision(B)': 0.72091, 'metrics/recall(B)': 0.97336, 'metrics/mAP50(B)': 0.86256, 'metrics/mAP50-95(B)': 0.8447, 'val/box_loss': 0.0065, 'val/cls_loss': 0.50605, 'val/dfl_loss': 0.78559, 'fitness': 0.84648}
Tuner: Best fitness model is /content/drive/MyDrive/codeit_level1_project/YOLOv8_epoch100_hyperparameters/yolo_tuning_results/train
Tuner: Best fitness hyperparameters are printed below.

Printing '/content/drive/MyDrive/codeit_level1_projec

## **5. YOLOv8x 모델 훈련 및 검증**

### 5.1 모델 구성 및 학습

- 앞서 튜닝한 하이퍼파라미터를 기반으로, YOLOv8x 모델을 활용하여 훈련을 진행하였다.
- 하이퍼파라미터는 model.tune()을 통해 YOLOv8n 모델로 탐색한 결과 중, 성능이 가장 우수했던 설정값을 적용하였다.
- 훈련은 총 100 epoch을 기준으로 하되, 5 epoch 동안 성능 개선이 없을 경우 자동 중단(early stopping) 되도록 설정하였다.

In [ ]:
model = YOLO("yolov8x.pt")

model.train(
    data='/content/drive/MyDrive/codeit_level1_project/data/data.yaml',
    imgsz=640,
    batch=16,
    workers=4,                 # 데이터 로딩 멀티스레딩
    patience=5,                # early stopping: 개선 없으면 조기 종료
    optimizer='AdamW',          # 옵티마이저
    seed=71,                   # 시드 고정으로 재현성 확보
    save=True,                 # 모델 체크포인트 저장 (기본 True)
    save_period=5,             # 매 5 에폭마다 저장
    project="/content/drive/MyDrive/codeit_level1_project/YOLOv8_epoch100_hyperparameters/pill_yolo_exp",   # 프로젝트 폴더명
    name="yolov8x_hyp0534_ep100",       # 실험 이름
    verbose=True,               # 학습 로그 출력
    lr0=0.0092,
    lrf=0.01,
    momentum=0.76932,
    weight_decay=0.0005,
    warmup_epochs=2.67276,
    warmup_momentum=0.74631,
    box=0.14182,
    cls=0.38274,
    hsv_h=0.01582,
    hsv_s=0.73822,
    hsv_v=0.28024,
    fliplr=0.55256,
    mosaic=1.0
)

Ultralytics 8.3.146 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (NVIDIA A100-SXM4-40GB, 40507MiB)
engine/trainer: agnostic_nms=False, amp=True, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=0.14182, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.38274, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/drive/MyDrive/codeit_level1_project/data/data.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, epochs=100, erasing=0.4, exist_ok=False, fliplr=0.55256, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.01582, hsv_s=0.73822, hsv_v=0.28024, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.0092, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8x.pt, momentum=0.76932, mosaic=1.0, multi_scale=False, name=yolov8x_hyp0534_ep100, nbs=64, nms=False, opset=None, optimize=False, optimizer=Ad

100%|██████████| 22.2M/22.2M [00:00<00:00, 41.6MB/s]

Overriding model.yaml nc=80 with nc=73

                   from  n    params  module                                       arguments                     
  0                  -1  1      2320  ultralytics.nn.modules.conv.Conv             [3, 80, 3, 2]                 
  1                  -1  1    115520  ultralytics.nn.modules.conv.Conv             [80, 160, 3, 2]               
  2                  -1  3    436800  ultralytics.nn.modules.block.C2f             [160, 160, 3, True]           
  3                  -1  1    461440  ultralytics.nn.modules.conv.Conv             [160, 320, 3, 2]              
  4                  -1  6   3281920  ultralytics.nn.modules.block.C2f             [320, 320, 6, True]           
  5                  -1  1   1844480  ultralytics.nn.modules.conv.Conv             [320, 640, 3, 2]              


  6                  -1  6  13117440  ultralytics.nn.modules.block.C2f             [640, 640, 6, True]           
  7                  -1  1   3687680  ultralytics.nn.modules.conv.Conv             [640, 640, 3, 2]              
  8                  -1  3   6969600  ultralytics.nn.modules.block.C2f             [640, 640, 3, True]           
  9                  -1  1   1025920  ultralytics.nn.modules.block.SPPF            [640, 640, 5]                 
 10                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 11             [-1, 6]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 12                  -1  3   7379200  ultralytics.nn.modules.block.C2f             [1280, 640, 3]                
 13                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 14             [-1, 4]  1         0  ultralytics.nn.modules.conv.Concat           [1]  

100%|██████████| 5.35M/5.35M [00:00<00:00, 314MB/s]


AMP: checks passed ✅
train: Fast image access ✅ (ping: 0.4±0.1 ms, read: 1.9±0.6 MB/s, size: 1723.5 KB)


train: Scanning /content/drive/MyDrive/codeit_level1_project/data/labels/train.cache... 1192 images, 0 backgrounds, 0 corrupt: 100%|██████████| 1192/1192 [00:00<?, ?it/s]


albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01, method='weighted_average', num_output_channels=3), CLAHE(p=0.01, clip_limit=(1.0, 4.0), tile_grid_size=(8, 8))
val: Fast image access ✅ (ping: 0.4±0.0 ms, read: 1.9±0.5 MB/s, size: 1737.6 KB)


val: Scanning /content/drive/MyDrive/codeit_level1_project/data/labels/val.cache... 297 images, 0 backgrounds, 2 corrupt: 100%|██████████| 297/297 [00:00<?, ?it/s]

val: /content/drive/MyDrive/codeit_level1_project/data/images/val/K-003351-016262-018357_0_2_0_2_75_000_200.png: ignoring corrupt image/label: non-normalized or out of bounds coordinates [     6.8878]
val: /content/drive/MyDrive/codeit_level1_project/data/images/val/K-003544-004543-012247-016551_0_2_0_2_70_000_200.png: ignoring corrupt image/label: non-normalized or out of bounds coordinates [     7.0293]


Plotting labels to /content/drive/MyDrive/codeit_level1_project/YOLOv8_epoch100_hyperparameters/pill_yolo_exp/yolov8x_hyp0534_ep100/labels.jpg... 
optimizer: AdamW(lr=0.0092, momentum=0.76932) with parameter groups 97 weight(decay=0.0), 104 weight(decay=0.0005), 103 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 4 dataloader workers
Logging results to /content/drive/MyDrive/codeit_level1_project/YOLOv8_epoch100_hyperparameters/pill_yolo_exp/yolov8x_hyp0534_ep100
Starting training for 100 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      1/100      11.7G    0.02864      3.964      1.733        101        640:   5%|▌         | 4/75 [00:32<05:23,  4.56s/it]

      1/100      11.7G    0.02887      3.957      1.745         95        640:   7%|▋         | 5/75 [00:34<04:06,  3.52s/it]
100%|██████████| 755k/755k [00:00<00:00, 91.8MB/s]
      1/100      11.8G     0.0285      3.744      1.756         42        640: 100%|██████████| 75/75 [01:29<00:00,  1.19s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:16<00:00,  1.64s/it]

                   all        295        890          0          0          0          0



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      2/100      13.3G    0.02144      2.671      1.511         49        640: 100%|██████████| 75/75 [00:18<00:00,  3.95it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:11<00:00,  1.10s/it]

                   all        295        890    0.00024    0.00562   0.000125   2.48e-05



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      3/100      13.3G    0.02116      2.525      1.495         60        640: 100%|██████████| 75/75 [00:25<00:00,  2.99it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:01<00:00,  5.08it/s]

                   all        295        890          0          0          0          0



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      4/100      13.3G     0.0159      2.208      1.285         42        640: 100%|██████████| 75/75 [00:23<00:00,  3.16it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.49it/s]


                   all        295        890      0.484      0.226      0.116     0.0996

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      5/100      13.4G    0.01466      1.951       1.22         45        640: 100%|██████████| 75/75 [00:25<00:00,  2.97it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.47it/s]

                   all        295        890      0.472        0.4      0.195      0.168



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      6/100      13.4G    0.01336      1.759      1.178         40        640: 100%|██████████| 75/75 [00:24<00:00,  3.02it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.89it/s]

                   all        295        890      0.375      0.302      0.195      0.171



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      7/100      13.4G    0.01198      1.546       1.12         42        640: 100%|██████████| 75/75 [00:24<00:00,  3.12it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.47it/s]


                   all        295        890      0.329      0.357      0.228      0.199

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      8/100      13.4G    0.01151      1.469      1.097         40        640: 100%|██████████| 75/75 [00:25<00:00,  2.95it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  3.80it/s]


                   all        295        890      0.404      0.428      0.272      0.236

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      9/100      13.4G    0.01112      1.385      1.094         32        640: 100%|██████████| 75/75 [00:25<00:00,  2.95it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.58it/s]


                   all        295        890      0.527      0.336      0.272      0.237

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     10/100      13.4G    0.01117        1.3      1.087         46        640: 100%|██████████| 75/75 [00:25<00:00,  2.95it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.63it/s]

                   all        295        890      0.382      0.582       0.38      0.344



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     11/100      13.4G    0.01089      1.241       1.07         52        640: 100%|██████████| 75/75 [00:25<00:00,  3.00it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.43it/s]


                   all        295        890       0.43      0.475      0.387      0.347

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     12/100      14.5G    0.01036      1.186      1.049         48        640: 100%|██████████| 75/75 [00:26<00:00,  2.85it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.52it/s]


                   all        295        890      0.407      0.572      0.407      0.369

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     13/100      14.5G    0.01016       1.14      1.054         44        640: 100%|██████████| 75/75 [00:23<00:00,  3.24it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.68it/s]

                   all        295        890      0.414      0.727      0.535       0.49



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     14/100      14.5G   0.009648      1.066      1.027         49        640: 100%|██████████| 75/75 [00:25<00:00,  2.96it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.81it/s]

                   all        295        890      0.362      0.654      0.507      0.474



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     15/100      14.5G   0.009778      1.042      1.044         51        640: 100%|██████████| 75/75 [00:22<00:00,  3.26it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.87it/s]

                   all        295        890      0.493       0.71      0.605      0.568



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     16/100      14.5G    0.00961     0.9939      1.022         58        640: 100%|██████████| 75/75 [00:25<00:00,  2.99it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.58it/s]

                   all        295        890      0.411      0.691      0.552      0.506



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     17/100      14.5G   0.009646     0.9773      1.021         46        640: 100%|██████████| 75/75 [00:23<00:00,  3.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.85it/s]

                   all        295        890       0.46      0.789      0.642      0.602



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     18/100      14.5G   0.009072     0.9324      1.007         44        640: 100%|██████████| 75/75 [00:17<00:00,  4.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.67it/s]

                   all        295        890      0.539       0.73      0.694      0.654



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     19/100      14.5G   0.009099     0.9107      1.008         31        640: 100%|██████████| 75/75 [00:24<00:00,  3.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.75it/s]

                   all        295        890      0.482      0.715      0.639      0.605



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     20/100      14.5G   0.008739     0.8669      0.987         49        640: 100%|██████████| 75/75 [00:24<00:00,  3.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.87it/s]

                   all        295        890      0.519      0.759      0.702       0.66



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     21/100      14.5G   0.008775     0.8915     0.9987         30        640: 100%|██████████| 75/75 [00:25<00:00,  2.99it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.68it/s]

                   all        295        890      0.464      0.834      0.678       0.63



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     22/100      14.5G    0.00909     0.8679      1.011         46        640: 100%|██████████| 75/75 [00:18<00:00,  4.16it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.82it/s]

                   all        295        890      0.506      0.839      0.657       0.62



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     23/100      14.5G   0.008823     0.8308     0.9957         54        640: 100%|██████████| 75/75 [00:23<00:00,  3.14it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.56it/s]

                   all        295        890      0.408      0.738      0.636      0.598



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     24/100      14.5G   0.008688     0.8032     0.9922         42        640: 100%|██████████| 75/75 [00:23<00:00,  3.14it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.77it/s]

                   all        295        890      0.489      0.774      0.674      0.636



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     25/100      14.5G   0.008464     0.7966     0.9863         41        640: 100%|██████████| 75/75 [00:19<00:00,  3.83it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.59it/s]

                   all        295        890      0.549      0.818      0.721      0.678



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     26/100      14.5G   0.008384     0.7839      0.984         55        640: 100%|██████████| 75/75 [00:24<00:00,  3.02it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.61it/s]

                   all        295        890      0.573      0.743      0.698      0.665



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     27/100      14.5G   0.008346     0.7849     0.9834         51        640: 100%|██████████| 75/75 [00:24<00:00,  3.05it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.70it/s]

                   all        295        890      0.531      0.807      0.738      0.701



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     28/100      14.5G   0.008354      0.771     0.9836         39        640: 100%|██████████| 75/75 [00:18<00:00,  4.14it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:01<00:00,  5.07it/s]

                   all        295        890      0.541      0.848      0.728        0.7



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     29/100      14.5G   0.008234      0.737     0.9765         43        640: 100%|██████████| 75/75 [00:16<00:00,  4.42it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.93it/s]

                   all        295        890        0.7       0.77      0.788      0.755



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     30/100      14.5G    0.00826     0.7178     0.9754         55        640: 100%|██████████| 75/75 [00:17<00:00,  4.24it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.76it/s]

                   all        295        890       0.54      0.783      0.723      0.663



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     31/100      14.5G   0.008255     0.7258     0.9727         47        640: 100%|██████████| 75/75 [00:16<00:00,  4.52it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.87it/s]

                   all        295        890      0.584      0.724      0.694      0.656



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     32/100      14.5G   0.008091     0.6885     0.9772         51        640: 100%|██████████| 75/75 [00:18<00:00,  4.16it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.98it/s]

                   all        295        890      0.552      0.851      0.767      0.727



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     33/100      14.5G   0.007816     0.6713     0.9621         47        640: 100%|██████████| 75/75 [00:16<00:00,  4.46it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.72it/s]

                   all        295        890      0.584       0.91        0.8      0.763



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     34/100      14.5G    0.00791     0.6767     0.9711         56        640: 100%|██████████| 75/75 [00:24<00:00,  3.05it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.76it/s]

                   all        295        890      0.558      0.894       0.74      0.709



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     35/100      14.5G   0.008004     0.6916     0.9758         48        640: 100%|██████████| 75/75 [00:22<00:00,  3.27it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.63it/s]

                   all        295        890      0.602      0.881      0.773      0.743



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     36/100      14.5G   0.007812     0.6724     0.9608         49        640: 100%|██████████| 75/75 [00:16<00:00,  4.48it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.95it/s]

                   all        295        890      0.624      0.809      0.765      0.724



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     37/100      14.5G   0.007872     0.6758     0.9644         31        640: 100%|██████████| 75/75 [00:16<00:00,  4.47it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.90it/s]

                   all        295        890      0.675      0.885      0.822      0.792



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     38/100      14.5G   0.007708     0.6561     0.9634         42        640: 100%|██████████| 75/75 [00:17<00:00,  4.37it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:01<00:00,  5.35it/s]

                   all        295        890       0.64      0.915      0.806      0.775



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     39/100      14.5G   0.007699     0.6591     0.9617         42        640: 100%|██████████| 75/75 [00:22<00:00,  3.27it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.48it/s]

                   all        295        890      0.664       0.86      0.819      0.784



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     40/100      14.5G    0.00755     0.6297     0.9544         58        640: 100%|██████████| 75/75 [00:23<00:00,  3.16it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.62it/s]

                   all        295        890      0.602      0.905      0.796      0.753



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     41/100      14.6G    0.00736     0.6153     0.9458         46        640: 100%|██████████| 75/75 [00:23<00:00,  3.16it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.64it/s]

                   all        295        890      0.666      0.895      0.822      0.796



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     42/100      14.6G   0.007511     0.6285     0.9499         46        640: 100%|██████████| 75/75 [00:25<00:00,  2.92it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.95it/s]

                   all        295        890      0.644      0.863      0.824       0.79



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     43/100      14.6G   0.007489     0.6173     0.9499         50        640: 100%|██████████| 75/75 [00:16<00:00,  4.56it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.92it/s]

                   all        295        890        0.7       0.89       0.82      0.789



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     44/100      14.6G   0.007376     0.6144     0.9526         63        640: 100%|██████████| 75/75 [00:16<00:00,  4.49it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.67it/s]

                   all        295        890      0.689      0.919       0.82      0.792



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     45/100      14.6G   0.007346        0.6     0.9472         54        640: 100%|██████████| 75/75 [00:16<00:00,  4.49it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.84it/s]

                   all        295        890      0.621      0.966      0.821      0.791



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     46/100      14.6G   0.007356     0.5928     0.9513         57        640: 100%|██████████| 75/75 [00:16<00:00,  4.47it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:01<00:00,  5.03it/s]

                   all        295        890      0.677      0.928      0.837       0.81



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     47/100      14.6G   0.007355     0.5862     0.9486         46        640: 100%|██████████| 75/75 [00:19<00:00,  3.94it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:01<00:00,  5.04it/s]

                   all        295        890      0.726      0.884      0.841      0.817



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     48/100      14.6G    0.00719     0.5753     0.9436         43        640: 100%|██████████| 75/75 [00:18<00:00,  4.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.60it/s]

                   all        295        890       0.71      0.953      0.851      0.826



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     49/100      14.6G   0.007313     0.5758     0.9449         39        640: 100%|██████████| 75/75 [00:25<00:00,  3.00it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.96it/s]

                   all        295        890      0.692      0.924      0.824        0.8



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     50/100      14.6G    0.00692     0.5498     0.9324         49        640: 100%|██████████| 75/75 [00:23<00:00,  3.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.71it/s]

                   all        295        890      0.699      0.904      0.826      0.801



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     51/100      14.6G   0.007198     0.5616     0.9443         48        640: 100%|██████████| 75/75 [00:23<00:00,  3.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.66it/s]

                   all        295        890      0.622      0.899      0.801      0.772



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     52/100      14.6G   0.007251     0.5826     0.9425         38        640: 100%|██████████| 75/75 [00:24<00:00,  3.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  4.65it/s]

                   all        295        890      0.684       0.94      0.826      0.805



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     53/100      14.6G   0.006832     0.5412      0.927         41        640: 100%|██████████| 75/75 [00:16<00:00,  4.43it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:01<00:00,  5.06it/s]

                   all        295        890      0.698      0.934      0.834      0.807
EarlyStopping: Training stopped early as no improvement observed in last 5 epochs. Best results observed at epoch 48, best model saved as best.pt.
To update EarlyStopping(patience=5) pass a new patience value, i.e. `patience=300` or use `patience=0` to disable EarlyStopping.



53 epochs completed in 0.414 hours.
Optimizer stripped from /content/drive/MyDrive/codeit_level1_project/YOLOv8_epoch100_hyperparameters/pill_yolo_exp/yolov8x_hyp0534_ep100/weights/last.pt, 136.8MB
Optimizer stripped from /content/drive/MyDrive/codeit_level1_project/YOLOv8_epoch100_hyperparameters/pill_yolo_exp/yolov8x_hyp0534_ep100/weights/best.pt, 136.8MB

Validating /content/drive/MyDrive/codeit_level1_project/YOLOv8_epoch100_hyperparameters/pill_yolo_exp/yolov8x_hyp0534_ep100/weights/best.pt...
Ultralytics 8.3.146 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (NVIDIA A100-SXM4-40GB, 40507MiB)
Model summary (fused): 112 layers, 68,193,867 parameters, 0 gradients, 257.8 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 10/10 [00:02<00:00,  3.35it/s]


                   all        295        890      0.711      0.953      0.852      0.825
            보령부스파정 5mg         24         24      0.669          1      0.919      0.909
           뮤테란캡슐 100mg          1          1      0.811          1      0.995      0.995
           일양하이트린정 2mg          2          2       0.45          1      0.497      0.497
    기넥신에프정(은행엽엑스)(수출용)         90         90      0.728          1      0.777      0.771
     무코스타정(레바미피드)(비매품)          5          5      0.628          1      0.862      0.829
                  알드린정         10         10      0.487          1      0.647      0.647
          뉴로메드정(옥시라세탐)         10         10      0.794          1      0.889      0.841
            타이레놀정500mg          4          4          1      0.447        0.7      0.644
          에어탈정(아세클로페낙)         11         11      0.892      0.818      0.931      0.885
         삼남건조수산화알루미늄겔정         35         35      0.796          1       0.87      0.859
타이레놀이알서방정(아세트아미노펜)(수출

ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x7bf50db1a650>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0

### 5.2 검증 데이터 성능 평가
- 훈련 완료 후, 검증 데이터셋(295장, 890개 객체)을 대상으로 모델 성능을 평가하였다.

- 평가에는 튜닝된 모델의 best weight(best.pt)를 로드하여 model.val() 함수를 사용하였다.

- 검증 과정에서 일부 라벨이 비정상 범위를 가지는 이미지 2장은 자동으로 무시되었다.

| 지표                | 값     | 설명                         |
| ----------------- | ----- | -------------------------- |
| **Precision**     | 0.710 | 예측한 것 중 정답인 비율             |
| **Recall**        | 0.953 | 실제 객체 중 예측에 성공한 비율         |
| **mAP\@0.5**      | 0.852 | IoU ≥ 0.5일 때의 평균 정밀도       |
| **mAP\@0.5:0.95** | 0.827 | IoU 0.5\~0.95 범위에서의 평균 정밀도 |


In [ ]:
model = YOLO("/content/drive/MyDrive/codeit_level1_project/YOLOv8_epoch100_hyperparameters/pill_yolo_exp/yolov8x_hyp0534_ep100/weights/best.pt")
metrics = model.val()
print(metrics.box)

Ultralytics 8.3.146 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (NVIDIA A100-SXM4-40GB, 40507MiB)
Model summary (fused): 112 layers, 68,193,867 parameters, 0 gradients, 257.8 GFLOPs


100%|██████████| 22.2M/22.2M [00:00<00:00, 136MB/s] 


val: Fast image access ✅ (ping: 0.4±0.1 ms, read: 85.8±181.1 MB/s, size: 1690.6 KB)


val: Scanning /content/drive/MyDrive/codeit_level1_project/data/labels/val.cache... 297 images, 0 backgrounds, 2 corrupt: 100%|██████████| 297/297 [00:00<?, ?it/s]

val: /content/drive/MyDrive/codeit_level1_project/data/images/val/K-003351-016262-018357_0_2_0_2_75_000_200.png: ignoring corrupt image/label: non-normalized or out of bounds coordinates [     6.8878]
val: /content/drive/MyDrive/codeit_level1_project/data/images/val/K-003544-004543-012247-016551_0_2_0_2_70_000_200.png: ignoring corrupt image/label: non-normalized or out of bounds coordinates [     7.0293]



                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 19/19 [00:05<00:00,  3.80it/s]


                   all        295        890       0.71      0.953      0.852      0.827
            보령부스파정 5mg         24         24      0.668          1      0.916      0.909
           뮤테란캡슐 100mg          1          1      0.804          1      0.995      0.995
           일양하이트린정 2mg          2          2      0.447          1      0.497      0.497
    기넥신에프정(은행엽엑스)(수출용)         90         90      0.728          1      0.777      0.771
     무코스타정(레바미피드)(비매품)          5          5      0.627          1      0.862      0.829
                  알드린정         10         10      0.487          1      0.647      0.647
          뉴로메드정(옥시라세탐)         10         10      0.794          1      0.889      0.849
            타이레놀정500mg          4          4          1       0.45      0.699      0.643
          에어탈정(아세클로페낙)         11         11      0.885      0.818      0.931      0.872
         삼남건조수산화알루미늄겔정         35         35      0.796          1       0.87      0.859
타이레놀이알서방정(아세트아미노펜)(수출

## **6. 모델 평가**
- 학습된 YOLOv8x 모델의 최적 가중치(best.pt)를 활용하여, 테스트 이미지에 대한 객체 탐지 예측을 수행하였다.

- 예측에는 Ultralytics의 model.predict() 함수를 사용하였으며, confidence threshold를 0.05로 낮춰 검출 누락 가능성을 줄이는 방식으로 설정하였다.

In [ ]:
test_results = model.predict(
    source=test_dir,
    conf=0.05,
    imgsz=640,
    save=False,
    save_txt=False,
    name="predict_kaggle"
)


image 1/843 /content/drive/MyDrive/codeit_level1_project/data/test_images/1.png: 640x512 1 뉴로메드정(옥시라세탐), 1 삼남건조수산화알루미늄겔정, 1 리리카캡슐 150mg, 1 세비카정 10/40mg, 58.1ms
image 2/843 /content/drive/MyDrive/codeit_level1_project/data/test_images/10.png: 640x512 2 보령부스파정 5mgs, 1 타이레놀이알서방정(아세트아미노펜)(수출용), 1 조인스정 200mg, 1 리리카캡슐 150mg, 12.1ms
image 3/843 /content/drive/MyDrive/codeit_level1_project/data/test_images/100.png: 640x512 2 보령부스파정 5mgs, 1 삐콤씨에프정 618.6mg/병, 1 리리카캡슐 150mg, 1 엑스포지정 5/160mg, 13.1ms
image 4/843 /content/drive/MyDrive/codeit_level1_project/data/test_images/1003.png: 640x512 1 기넥신에프정(은행엽엑스)(수출용), 1 리피토정 20mg, 1 비모보정 500/20mg, 1 로수바미브정 10/20mg, 12.1ms
image 5/843 /content/drive/MyDrive/codeit_level1_project/data/test_images/1004.png: 640x512 1 기넥신에프정(은행엽엑스)(수출용), 1 리피토정 20mg, 1 비모보정 500/20mg, 1 로수바미브정 10/20mg, 12.4ms
image 6/843 /content/drive/MyDrive/codeit_level1_project/data/test_images/1005.png: 640x512 1 기넥신에프정(은행엽엑스)(수출용), 1 리피토정 20mg, 1 비모보정 500/20mg, 1 로수바미브정 10/20mg, 12.2ms

## **7. 제출 파일 생성**

- 모델의 예측 결과를 기반으로 Kaggle 제출 포맷에 맞는 csv 파일을 생성한다.
- test_images 폴더의 이미지 파일명은 중간중간 누락된 번호가 많기 때문에, 단순한 순번(1, 2, 3...)이 아닌, 파일명에서 추출한 실제 숫자 ID를 그대로 사용해야 한다.
- 클래스 번호는 제출 시 반드시 주어진 category_id와 정확히 매핑해야 평가가 올바르게 이루어진다.

In [ ]:
# 클래스 매핑 로드
with open("/content/drive/MyDrive/codeit_level1_project/data/classid_to_categoryid.json", "r") as f:
    class_to_category = json.load(f)
    class_to_category = {int(k): v for k, v in class_to_category.items()}

# 결과 파싱 및 저장
rows = []
annotation_id = 1

for result in test_results:
    # image_id: 원본 이미지 경로에서 파일명 숫자 추출
    filename = result.path.split("/")[-1]
    image_id = int(filename.split(".")[0])  # 예: "0089.png" → 89

    boxes = result.boxes
    if boxes is None or boxes.xywh is None:
        continue

    for i in range(len(boxes.xywh)):
        x, y, w, h = boxes.xywh[i].tolist()
        score = float(boxes.conf[i])
        cls = int(boxes.cls[i])
        category_id = class_to_category.get(cls, -1)

        if category_id == -1:
            continue  # 매핑 실패한 클래스는 제외

        # COCO 포맷: x_min, y_min, width, height
        xmin = round(x - w / 2)
        ymin = round(y - h / 2)
        bbox_w = round(w)
        bbox_h = round(h)

        rows.append([
            annotation_id,
            image_id,
            category_id,
            xmin,
            ymin,
            bbox_w,
            bbox_h,
            round(score, 2)
        ])
        annotation_id += 1

# CSV 저장
df = pd.DataFrame(rows, columns=[
    "annotation_id", "image_id", "category_id",
    "bbox_x", "bbox_y", "bbox_w", "bbox_h", "score"
])
csv_path = "/content/submission.csv"
df.to_csv(csv_path, index=False)
print(f"CSV 저장 완료: {csv_path} / 총 {len(df)}개의 바운딩 박스")

CSV 저장 완료: /content/submission.csv / 총 3775개의 바운딩 박스
